streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java

```diff
@@ -504,6 +510,9 @@ public void advanceWallClockTime(final long advanceMs) {
             task.maybePunctuateSystemTime();
             task.commit();
         }
+        if (eosEnabled) {
```

Comment:
Why do we need this?

Reply:
@mjsax ,
During `eosEnabled = true` case `@test execution TopologyTestDriverTest#shouldPunctuateIfWallClockTimeAdvances`, `setup()`  gets called. During `TopologyTestDriver`  initialization in setup() method, we call 
1) `task.initializeTopology()`. And it calls `StreamTask#initializeTopology()`. If `eosEnabled == true` , we are setting `transactionInFlight = true`, in both places MockProducer#beginTransaction and StreamTask#initializeTopology(). It's redundant code.
I think removing `transactionInFlight = true;` from `StreamTask#initializeTopology()` should holds good. Please do let me know your inputs.
2) After testdriver setup done, `testDriver.advanceWallClockTime` gets called, and             `task.maybePunctuateSystemTime();` gets invoked and it calls `MockProducer#send()` . In `step:1` as `transactionInFlight is set to true`, Code: `this.uncommittedSends.add(record);` gets called and Record added to `uncommittedSends list`.
3). `TopologyTestDriver#advanceWallClockTime` Code: ` producer.commitTransaction();` gets called and it  calls `MockProducer#commitTransaction`. Code: `this.sent.addAll(this.uncommittedSends);` gets called and  adds `uncommittedSends list` to `sent list`.
4) `TopologyTestDriver#advanceWallClockTime` Code: `captureOutputRecords()` gets called, and in captureOutputRecords() method Code: ` output = producer.history();` gets called and MockProducer#history returns `sent list` done at `step:3`.
5) `TopologyTestDriverTest#shouldPunctuateIfWallClockTimeAdvances` Code: Line `OutputVerifier.compareKeyValue(...)` verifies output successfully, with the record captured during step4.

 If the  newly added code TopologyTestDriver#advanceWallClockTime `producer.commitTransaction();` is not present then, at Step:3, Record will still present in the `uncommittedSends list` and `Step:4` returns `null` and `step:5 fails`.
 
Please do let me know, if i am missing something.

Reply:
What you are saying makes sense. It exposes a bug in Kafka Streams itself. I created a JIRA for this: https://issues.apache.org/jira/browse/KAFKA-6906

I would recommend to fix the actual bug first -- committing the transaction in the test driver is not a proper fix but a work-around that masks the actual bug.

\cc @guozhangwang WDYT?

Reply:
@mjsax , Thank you for the input. Agreed work-around was not proper fix.. Will fix the bug.

Reply:
@mjsax , Can i exclude `TopologyTestDriverTest#shouldPunctuateIfWallClockTimeAdvances` when `eosEnabled` is `true` and remove  above code `producer.commitTransaction();`.
And fix this test case as part of : https://issues.apache.org/jira/browse/KAFKA-6906

Reply:
I would prefer to merge the fix first. There is still enough time till code freeze to merge both for 2.0 release. Let us know if this works for you -- if you insist, we can also follow your proposal.

Reply:
Sure @mjsax . We can merge the fix first. Thank you for accepting the PR.

Reply:
@mjsax , Once https://github.com/apache/kafka/pull/5105 gets merged, we shall remove this work around.